### PR TITLE
Use numba to speedup Angle operations

### PR DESCRIPTION
**Pull request by maxnoe**
This is a draft PR to start discussing about using numba for possible performance improvements in the coordinates code.

See also #13479 

### Description

Introduces numba jitted functions for critical code in `astropy.coordinates.angle`

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -32,6 +32,7 @@ install_requires =
     pyerfa>=2.0
     PyYAML>=3.13
     packaging>=19.0
+	numba>=0.55.2
```
Comment (pllim) [inline setup.cfg:35]: Indentation aside, this is quite a controversial proposal. Have you seen #12448  ? 😆 

```suggestion
    numba>=0.55.2
```

**Review by pllim — COMMENTED** [review]
If this PR is to go forward, it would need a change log and possibly maybe even what's new (but that can wait for now).

I am not sure how to test the numba part in CI.

**Comment by Cadair** [comment]
I generally think that numba, while a super cool and very useful project, isn't really the thing that a core package like astropy should be using. My reasoning for this can be summed up as it putting the complexity onto the user to get the performance advantage rather than the maintainers.

numba is a non-trivial runtime dependency, which means that it adds complexity to the *users* installing it and keeping their environment in such a way that it works (i.e. not updating Python too soon etc). Whereas tools like Cython / C-API extensions all put the complexity on the maintainers, because we have to worry about build environments, transpiling etc and then the users just install a binary which works (or if they are on a really exotic system have to have a working toolchain).

If we were to go this route, I would at least want to see us fall back to not using numba if it couldn't be imported.

**Comment by maxnoe** [comment]
>  My reasoning for this can be summed up as it putting the complexity onto the user to get the performance advantage rather than the maintainers.


But the user has no way to improve the performance of the astropy core functionanilty (other than monkeypatching, which won't work in many cases and is beyond hacky)

**Comment by maxnoe** [comment]
I agree, that numba is not the best approach here, which is why I also opened a cython based PR. 

But just shrugging off inefficient code in the core and putting that on users is not an approach we should take

**Comment by Cadair** [comment]
> But the user has no way to improve the performance of the astropy core functionanilty

I think you misunderstand me. It's our job as maintainers to improve the performance of the package, and we should choose the harder path for *us* if it makes our users lives easier. Us using numba makes our lives easier, but makes things more complex for our users. Us using Cython (etc) makes our lives slightly harder (as demonstrated by the nuance in the Cython PR) but is effectively transparent for our users.

**Comment by maxnoe** [comment]
Ah, sorry, yes, fully agree!

**Pull request closed without merging**